src/bt.rs

```diff
@@ -91,310 +97,296 @@ impl<V> Node<V> {
-    /// node at the end of the path. If the node is unreachable or already
-    /// contains a value, an error wrapping the value is returned. Otherwise,
-    /// a new node containing the value is inserted.
-    ///
-    /// # Returns
-    /// - `Ok(())` when the node is inserted at the end of the path.
-    /// - `Err(InsertNonEmptyNode(value))` when the tree already contains a
-    /// value at the end of the path.
-    /// - `Err(UnreachablePath(value))` when the end of the path is unreachable.
+    /// Gets a reference to the node located at the end of the path
+    /// traversing the tree from the specified node reference.
     ///
-    /// In the error cases, no insertion occurs and the value is returned to
-    /// the caller of this function.
-    pub fn insert(&mut self, path: &Path, value: V) -> Result<(), BinaryTreeError<V>> {
-        if let Some(node) = &mut self.root {
-            node.insert(path, value)
-        } else if path.is_empty() {
-            self.root = Some(Box::new(Node::new(value)));
-            Ok(())
-        } else {
-            Err(BinaryTreeError::UnreachableNode(value))
+    /// This function traverses the tree from the specified node until
+    /// reaching the node at the end of the path. It returns [None],
+    /// if the node is unreachable or nonexistent; otherwise, it returns
+    /// a reference to the node.
+    pub fn get_node_at(&self, node_ref: NodeRef, path: &Path) -> Option<NodeRef> {
+        self.is_valid_node_ref(node_ref).then_some(())?;
+
+        let new_index = self.nodes.len();
+        let mut node = (!self.nodes.is_empty()).then_some(node_ref.0);
+        for bit in path.iter() {
+            let next = &self.nodes[node?];
+            node = if !bit { next.left } else { next.right };
         }
-    }
 
-    /// Gets a reference to the value located at the end of the path.
-    ///
-    /// This function traverses the tree from the root node until reaching the
-    /// node at the end of the path. It returns [None], if the node is
-    /// unreachable or nonexistent. Otherwise, it returns a reference to the
-    /// value stored in the node.
-    pub fn get(&self, path: &Path) -> Option<&V> {
-        self.root.as_ref().and_then(|node| node.get(path))
+        node.map(NodeRef)
     }
 
-    /// Gets a mutable reference to the node located at the end of the path.
-    ///
-    /// This function traverses the tree from the root node until reaching the
-    /// node at the end of the path. It returns [None], if the node is
-    /// unreachable or nonexistent. Otherwise, it returns a mutable reference
-    /// to the node.
-    pub fn get_node(&mut self, path: &Path) -> Option<&mut Node<V>> {
-        self.root.as_mut().and_then(|node| node.get_node(path))
+    /// Checks whether the node reference is valid with respect to the tree.
+    fn is_valid_node_ref(&self, node_ref: NodeRef) -> bool {
+        !self.nodes.is_empty() && node_ref.0 < self.nodes.len()
```

Comment:
nit: I think the `!self.nodes.is_empty()` check is unnecessary: the `node_ref.0` is a `usize` and therefore nonnegative, so if `nodes` is empty, `node_ref.0 < self.nodes.len()` will be equivalent to `node_ref.0 < 0`, which is always false.